swarm/src/executor.rs

```diff
@@ -0,0 +1,48 @@
+use futures::executor::ThreadPool;
+use std::{future::Future, pin::Pin};
+
+/// Implemented on objects that can run a `Future` in the background.
+///
+/// > **Note**: While it may be tempting to implement this trait on types such as
+/// >           [`futures::stream::FuturesUnordered`], please note that passing an `Executor` is
+/// >           optional, and that `FuturesUnordered` (or a similar struct) will automatically
+/// >           be used as fallback by libp2p. The `Executor` trait should therefore only be
+/// >           about running `Future`s on a separate task.
+pub trait Executor {
+    /// Run the given future in the background until it ends.
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>);
+}
+
+impl<F: Fn(Pin<Box<dyn Future<Output = ()> + Send>>)> Executor for F {
+    fn exec(&self, f: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        self(f)
+    }
+}
+
+impl Executor for ThreadPool {
+    fn exec(&self, future: Pin<Box<dyn Future<Output = ()> + Send>>) {
+        self.spawn_ok(future)
+    }
+}
+
+#[cfg(feature = "tokio")]
+#[derive(Default, Debug, Clone, Copy)]
+pub(crate) struct TokioExecutor;
```

Comment:
Why not expose `TokioExecutor` and `AsyncStdExecutor`? Folks that want to specify configs, e.g. `dial_concurrency_factor` need to use the `SwarmBuilder`. Instead of implementing `Executor` on their custom type wrapping e.g. the `async-std` executor, they could just reuse our implementation. Thoughts @umgefahren and @thomaseizinger?

Reply:
Yes, I think you are right. If Thomas agrees I can open a PR real quick changing that and adding some documentation.

Reply:
Alternatively, we could add the same set of `with_` APIs.

I am okay either way. (I thought we had two layers of `with_` APIs)

Reply:
Can we just get rid of `SwarmBuilder` instead maybe? Is there a reason why any of the config values could not be set on `Swarm` itself?

Reply:
Trying to upgrade iroh to laster `master` and blocked on this :/

Reply:
I nedd the SwarmBuilder and have to reimplement the Executor trait by copy pasting the one in here, to use tokio.

Reply:
> Can we just get rid of `SwarmBuilder` instead maybe? Is there a reason why any of the config values could not be set on `Swarm` itself?

@umgefahren Would you mind:

- sending a hotfix PR that makes these types public
- attempting to remove `SwarmBuilder` altogether by moving the current "builder" functions directly onto `Swarm`

LMK if you are not keen for the latter, then I'll try myself :)

Reply:
Yes, I can do that (although I'm done for today)

Reply:
Done #3155